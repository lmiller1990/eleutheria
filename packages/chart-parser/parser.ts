export interface BaseNote {
  id: string;
  column: number;
  char: string;
  ms: number;
}

export interface ChartMetadata {
  title: string;
  bpm: number;
  offset: number;
}

export interface ParsedChart {
  metadata: ChartMetadata;
  notes: BaseNote[];
}

type Measure = string[];

const validQuantizations = [4, 8, 12, 16] as const;

type ValidQuantization = typeof validQuantizations[number];

function measureQuantizationValid(measure: Measure) {
  if (validQuantizations.includes(measure.length as ValidQuantization)) {
    return true;
  }
  throw Error(`${measure.length} is not a valid quanization`);
}

export function parseChart(
  dataJson: Record<string, string>,
  chartRaw: string
): ParsedChart {
  const lines = chartRaw.split("\n").filter((x) => x.trim().length > 0);

  const measures = lines.reduce<{
    measures: Measure[];
    currMeasure: Measure;
  }>(
    (acc, _curr, idx) => {
      const curr = _curr.trim();

      if (curr === "," && measureQuantizationValid(acc.currMeasure)) {
        acc.measures.push(acc.currMeasure);
        acc.currMeasure = [];
        return acc;
      }

      acc.currMeasure.push(curr);

      // end of chart
      if (idx + 1 === lines.length) {
        acc.measures.push(acc.currMeasure);
      }

      return acc;
    },
    { measures: [], currMeasure: [] }
  ).measures;

  const bpm = parseFloat(dataJson.bpm);

  const _4th = 60 / bpm;
  const _8th = _4th / 2;
  const _16th = _8th / 2;
  const measureMs = _4th * 4;

  const qMap = new Map<ValidQuantization, number>([
    [4, _4th],
    [8, _8th],
    [16, _16th],
  ]);

  const getQuantizationMs = (m: Measure) => {
    const q = qMap.get(m.length as ValidQuantization);
    if (!q) {
      throw Error(`${m.length} is not a valid quanization`);
    }
    return q;
  };

  const notes = measures.reduce<{
    notes: BaseNote[];
    measureCount: number;
    noteCount: number;
  }>(
    (acc, measure) => {
      const q = getQuantizationMs(measure);

      return {
        measureCount: acc.measureCount + 1,
        noteCount: acc.noteCount + measure.length,
        notes: [
          ...acc.notes,
          ...measure.reduce<BaseNote[]>((_notes, row, idx) => {
            if (row.split("").every((col) => col === "0")) {
              return _notes;
            }

            const _newNotes: BaseNote[] = [];

            const columns = row.split("");

            for (let i = 0; i < columns.length; ++i) {
              const col = columns[i];
              if (col === "0") {
                continue;
              }

              _newNotes.push({
                id: (acc.noteCount + idx + 1).toString(),
                char: col,
                column: i,
                ms: (acc.measureCount * measureMs + q * idx) * 1000,
              });
            }

            return _notes.concat(
              ..._newNotes.sort((x, y) => x.char.localeCompare(y.char))
            );
          }, []),
        ],
      };
    },
    { notes: [], measureCount: 0, noteCount: 0 }
  ).notes;

  return {
    metadata: {
      title: dataJson.title,
      bpm,
      offset: parseInt(dataJson.offset, 10),
    },
    notes,
  };
}
