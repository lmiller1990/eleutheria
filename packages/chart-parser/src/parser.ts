export interface BaseNote {
  id: string;
  column: number;
  char: string;
  ms: number;
  measureNumber: number;
}

export interface ChartMetadata {
  title: string;
  bpm: number;
  offset: number;
  artist: string;
  banner: string;
  charts: Array<{
    difficulty: string;
    level: number;
  }>;
  [x: string]: any;
}

export interface ParsedTapNoteChart {
  metadata: ChartMetadata;
  tapNotes: BaseNote[];
}

export type HoldNote = BaseNote[];

export interface ParsedHoldNoteChart {
  metadata: ChartMetadata;
  holdNotes: HoldNote[];
}

export type ValidQuantization = typeof validQuantizations[number];

export const validQuantizations = [4, 8, 12, 16, 24, 32] as const;

interface GetQuantization {
  quantization: ValidQuantization;
  measureMs: number;
}

export function getQuantizationMs(
  measure: string[],
  bpm: number
): GetQuantization {
  const _4th = 60 / bpm;
  const _8th = _4th / 2;
  const _12th = _4th / 3;
  const _16th = _8th / 2;
  const _24th = _12th / 2;
  const _32nd = _16th / 2;
  const measureMs = _4th * 4;

  const qMap = new Map<ValidQuantization, number>([
    [4, _4th],
    [8, _8th],
    [12, _12th],
    [16, _16th],
    [24, _24th],
    [32, _32nd],
  ]);

  const q = qMap.get(measure.length as ValidQuantization);
  if (!q) {
    throw Error(`${measure.length} is not a valid quanization`);
  }
  return {
    quantization: q as ValidQuantization,
    measureMs,
  };
}

type Measure = string[];

function measureQuantizationValid(measure: Measure) {
  if (validQuantizations.includes(measure.length as ValidQuantization)) {
    return true;
  }
  throw Error(`${measure.length} is not a valid quanization`);
}

export function parseChart(
  dataJson: ChartMetadata,
  chartRaw: string
): ParsedTapNoteChart {
  let i = 0;

  function incId() {
    i++;
    return i.toString();
  }
  const lines = chartRaw.split("\n").filter((x) => x.trim().length > 0);

  const measures = lines.reduce<{
    measures: Measure[];
    currMeasure: Measure;
  }>(
    (acc, _curr, idx) => {
      const curr = _curr.trim();

      if (curr.startsWith(",") && measureQuantizationValid(acc.currMeasure)) {
        acc.measures.push(acc.currMeasure);
        acc.currMeasure = [];
        return acc;
      }

      acc.currMeasure.push(curr);

      // end of chart
      if (idx + 1 === lines.length) {
        acc.measures.push(acc.currMeasure);
      }

      return acc;
    },
    { measures: [], currMeasure: [] }
  ).measures;

  const tapNotes = measures.reduce<{
    tapNotes: BaseNote[];
    measureCount: number;
    noteCount: number;
  }>(
    (acc, measure) => {
      const { quantization, measureMs } = getQuantizationMs(
        measure,
        dataJson.bpm
      );

      return {
        measureCount: acc.measureCount + 1,
        noteCount: acc.noteCount + measure.length,
        tapNotes: [
          ...acc.tapNotes,
          ...measure.reduce<BaseNote[]>((_notes, row, idx) => {
            if (row.split("").every((col) => col === "0")) {
              return _notes;
            }

            const _newNotes: BaseNote[] = [];

            const columns = row.split("");

            for (let i = 0; i < columns.length; ++i) {
              const col = columns[i];
              if (col === "0") {
                continue;
              }

              _newNotes.push({
                id: incId(),
                char: col,
                column: i,
                measureNumber: acc.measureCount,
                ms: (acc.measureCount * measureMs + quantization * idx) * 1000,
              });
            }

            return _notes.concat(
              ..._newNotes.sort((x, y) => x.char.localeCompare(y.char))
            );
          }, []),
        ],
      };
    },
    { tapNotes: [], measureCount: 0, noteCount: 0 }
  ).tapNotes;

  return {
    metadata: {
      title: dataJson.title,
      bpm: dataJson.bpm,
      banner: dataJson.banner,
      offset: dataJson.offset,
      charts: dataJson.charts,
      artist: dataJson.artist,
    },
    tapNotes,
  };
}

// uses parseChart then derives holds.
export function parseHoldsChart(
  dataJson: ChartMetadata,
  chartRaw: string
): ParsedHoldNoteChart {
  const result = parseChart(dataJson, chartRaw);

  let holds: HoldNote[] = [];
  let holdMap = new Map<number, HoldNote>();

  for (const note of result.tapNotes) {
    if (note.char === "1") {
      // start of new hold.
      if (holdMap.has(note.column)) {
        holds.push(holdMap.get(note.column)!);
      }
      holdMap.set(note.column, [{ ...note, id: `h${note.id}` }]);
    } else {
      const curr = holdMap.get(note.column)!;
      holdMap.set(note.column, [...curr, { ...note, id: `h${note.id}` }]);
    }
  }

  for (const [_k, hold] of holdMap) {
    holds.push(hold);
  }

  holds = holds.map((hold) =>
    hold.sort((x, y) => parseInt(x.char, 10) - parseInt(y.char, 10))
  );

  return {
    ...result,
    holdNotes: holds,
  };
}
