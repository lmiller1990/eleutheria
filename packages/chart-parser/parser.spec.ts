import fs from "fs";
import { describe, it } from "vitest";
import path from "path";
import { ChartMetadata, parseChart } from "./parser";

function loadData() {
  const exampleDir = path.resolve(
    __dirname,
    "test",
    "resources",
    "example-song"
  );
  const metadata = fs.readFileSync(path.join(exampleDir, "data.json"), "utf-8");
  const chart = fs.readFileSync(
    path.join(exampleDir, "example-song.chart"),
    "utf-8"
  );

  return {
    metadata: JSON.parse(metadata),
    chart,
  };
}

describe("parseChart", () => {
  it("works", () => {
    const data = loadData();
    const actual = parseChart(data.metadata, data.chart);

    console.log(actual)

    for (let i = 0; i < actual.notes.length; ++i) {
      let c = actual.notes[i]
      let n = actual.notes[i + 1]

      if (c && n) {
        console.log(n.ms - c.ms)
      }
    }
  });
});
