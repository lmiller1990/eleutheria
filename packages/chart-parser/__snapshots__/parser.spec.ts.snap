// Vitest Snapshot v1

exports[`parseChart > works 1`] = `
{
  "metadata": {
    "bpm": 120,
    "offset": 0,
    "title": "Example Song",
  },
  "tapNotes": [
    {
      "char": "N",
      "column": 0,
      "id": "1",
      "ms": 0,
    },
    {
      "char": "N",
      "column": 3,
      "id": "2",
      "ms": 500,
    },
    {
      "char": "N",
      "column": 1,
      "id": "3",
      "ms": 1000,
    },
    {
      "char": "N",
      "column": 2,
      "id": "4",
      "ms": 1500,
    },
    {
      "char": "N",
      "column": 0,
      "id": "5",
      "ms": 2000,
    },
    {
      "char": "N",
      "column": 3,
      "id": "6",
      "ms": 2250,
    },
    {
      "char": "N",
      "column": 0,
      "id": "7",
      "ms": 2500,
    },
    {
      "char": "N",
      "column": 3,
      "id": "8",
      "ms": 2750,
    },
    {
      "char": "N",
      "column": 0,
      "id": "9",
      "ms": 3000,
    },
    {
      "char": "N",
      "column": 1,
      "id": "10",
      "ms": 3000,
    },
    {
      "char": "N",
      "column": 2,
      "id": "11",
      "ms": 3250,
    },
    {
      "char": "N",
      "column": 3,
      "id": "12",
      "ms": 3250,
    },
    {
      "char": "N",
      "column": 0,
      "id": "13",
      "ms": 3500,
    },
    {
      "char": "N",
      "column": 3,
      "id": "14",
      "ms": 3500,
    },
    {
      "char": "N",
      "column": 1,
      "id": "15",
      "ms": 3750,
    },
    {
      "char": "N",
      "column": 2,
      "id": "16",
      "ms": 3750,
    },
  ],
}
`;

exports[`parseHoldsChart > supports multiple holds holds-simultaneously 1`] = `
{
  "holdNotes": [
    [
      {
        "char": "1",
        "column": 0,
        "id": "h1",
        "ms": 0,
      },
      {
        "char": "2",
        "column": 0,
        "id": "h3",
        "ms": 1000,
      },
    ],
    [
      {
        "char": "1",
        "column": 5,
        "id": "h2",
        "ms": 0,
      },
      {
        "char": "2",
        "column": 5,
        "id": "h4",
        "ms": 1000,
      },
    ],
  ],
  "metadata": {
    "bpm": 120,
    "offset": 0,
    "title": "Example Song",
  },
  "tapNotes": [
    {
      "char": "1",
      "column": 0,
      "id": "1",
      "ms": 0,
    },
    {
      "char": "1",
      "column": 5,
      "id": "2",
      "ms": 0,
    },
    {
      "char": "2",
      "column": 0,
      "id": "3",
      "ms": 1000,
    },
    {
      "char": "2",
      "column": 5,
      "id": "4",
      "ms": 1000,
    },
  ],
}
`;

exports[`parseHoldsChart > works 1`] = `
{
  "holdNotes": [
    [
      {
        "char": "1",
        "column": 3,
        "id": "h3",
        "ms": 2000,
      },
      {
        "char": "2",
        "column": 3,
        "id": "h4",
        "ms": 2750,
      },
    ],
    [
      {
        "char": "1",
        "column": 0,
        "id": "h1",
        "ms": 0,
      },
      {
        "char": "2",
        "column": 0,
        "id": "h2",
        "ms": 1500,
      },
    ],
    [
      {
        "char": "1",
        "column": 3,
        "id": "h5",
        "ms": 3000,
      },
      {
        "char": "2",
        "column": 3,
        "id": "h6",
        "ms": 3750,
      },
    ],
  ],
  "metadata": {
    "bpm": 120,
    "offset": 0,
    "title": "Example Song",
  },
  "tapNotes": [
    {
      "char": "1",
      "column": 0,
      "id": "1",
      "ms": 0,
    },
    {
      "char": "2",
      "column": 0,
      "id": "2",
      "ms": 1500,
    },
    {
      "char": "1",
      "column": 3,
      "id": "3",
      "ms": 2000,
    },
    {
      "char": "2",
      "column": 3,
      "id": "4",
      "ms": 2750,
    },
    {
      "char": "1",
      "column": 3,
      "id": "5",
      "ms": 3000,
    },
    {
      "char": "2",
      "column": 3,
      "id": "6",
      "ms": 3750,
    },
  ],
}
`;
